### ISSUE TITLE
Cannot repeat FB15k tutorial results, and training doesn't converge (learning rate decay problem?)

### ISSUE THREAD

**Issue by dany-nonstop** [facebookresearch/PyTorch-BigGraph#145]
## Steps to reproduce

  1. in ubuntu 20.04 install python 3.8.2 and install default pytorch (1.5.0) and torchbiggraph (1.0.0)
  2. follow exactly the same code as in [README.md](https://github.com/facebookresearch/PyTorch-BigGraph/blob/master/README.md) to prepare data and train the model, then copy the code as in the document's [Downstream tasks](https://torchbiggraph.readthedocs.io/en/latest/downstream_tasks.html#ranking) to `rank.py`
  3. run `python rank.py` to search for the capital of entity France with freebase id `/m/0f8l9c`

## Observed Results

I trained twice with similar results
['/m/0f8l9c', '/m/05qtj', '/m/0cx2r', '/m/0lpp8', '/m/06c62']
they are France, Paris, Alsacia, Burgundy, Rome respectively. 
['/m/0f8l9c', '/m/05qtj', '/m/0dprg', '/m/0p0mx', '/m/0f0sbl']
they are France, Paris, Lyon, Picardy, Corsica respectively. 
Obviously they are not as good as the training results obtained in the document.

## Expected Results

according to the document, they should be Paris, Lyon, Martinique, Strasbourg and Rouen respectively. But I understand they may be due to some randomness in the seed, and I'm ok with it. The bigger problem is in the convergence of the embedding itself. From the training_stats.json it seems the loss jumps wildly even at epoch 43 on (from 64 to 495 then to 20), and actually volotators doesn't go any better (why?). It seems lr_decay parameter is 0 for RowAdagrad, is this the reason trainging doesn't converge?

``` 
"loss": 1903.6521253207418, "violators_lhs": 107.05835551452782, "violators_rhs": 102.91714651179156
"loss": 4292.073432349616, "violators_lhs": 11.41429641803031, "violators_rhs": 7.260012584291989
"loss": 1573.8243129643736, "violators_lhs": 9.476586593589463, "violators_rhs": 5.816225871482918
"loss": 1396.7156881958226, "violators_lhs": 8.894726602116975, "violators_rhs": 5.432233173683927
"loss": 177.45646651809275, "violators_lhs": 8.585467626494902, "violators_rhs": 5.238021120084778
"loss": 644.866844992925, "violators_lhs": 8.389461069416445, "violators_rhs": 5.128637129456764
"loss": 112.43614233419257, "violators_lhs": 8.285348820843561, "violators_rhs": 5.05228897508393
"loss": 276.7644875501308, "violators_lhs": 8.199823240372396, "violators_rhs": 4.997663212885652
"loss": 406.2410944849404, "violators_lhs": 8.133703962810106, "violators_rhs": 4.953616121140369
"loss": 807.4814888755087, "violators_lhs": 8.070660385559526, "violators_rhs": 4.924908205041168
"loss": 163.47298068334567, "violators_lhs": 8.041877957205129, "violators_rhs": 4.902138501724131
"loss": 262.1528420863316, "violators_lhs": 8.024891232805262, "violators_rhs": 4.873707936797049
"loss": 249.52263946642478, "violators_lhs": 8.007283572945427, "violators_rhs": 4.87017481402983
"loss": 235.30940522618616, "violators_lhs": 7.981345028997686, "violators_rhs": 4.842154066506327
"loss": 551.8503512859997, "violators_lhs": 7.980285299146007, "violators_rhs": 4.827620037173336
"loss": 212.33691929128392, "violators_lhs": 7.979142777899665, "violators_rhs": 4.798446419479159
"loss": 371.9241839196795, "violators_lhs": 7.960452620554619, "violators_rhs": 4.8028178051173365
"loss": 92.40596523845737, "violators_lhs": 7.9385066916144735, "violators_rhs": 4.7973804802728806
"loss": 279.3273347426459, "violators_lhs": 7.958155159352737, "violators_rhs": 4.787250953135931
"loss": 90.23751362826499, "violators_lhs": 7.9438902020523985, "violators_rhs": 4.775958206903974
"loss": 179.67307411351538, "violators_lhs": 7.943840527215602, "violators_rhs": 4.763179355137827
"loss": 55.53482347952957, "violators_lhs": 7.924163082489206, "violators_rhs": 4.743642655782358
"loss": 50.37758790416754, "violators_lhs": 7.951405590902882, "violators_rhs": 4.7498830571550394
"loss": 50.57534322063534, "violators_lhs": 7.937734661859246, "violators_rhs": 4.739600365937965
"loss": 60.549725050303856, "violators_lhs": 7.949921555153558, "violators_rhs": 4.744776897889233
"loss": 89.86806286613968, "violators_lhs": 7.94183076611017, "violators_rhs": 4.7297771669612665
"loss": 47.80738782428496, "violators_lhs": 7.9473612312736215, "violators_rhs": 4.739227804661984
"loss": 19.18719214287208, "violators_lhs": 7.944829884381817, "violators_rhs": 4.716789266923596
"loss": 43.602391319856494, "violators_lhs": 7.964985449412388, "violators_rhs": 4.724780706293388
"loss": 44.6012392490998, "violators_lhs": 7.949391690227718, "violators_rhs": 4.715402511063
"loss": 43.6270909492506, "violators_lhs": 7.9582545090263315, "violators_rhs": 4.706655600216913
"loss": 60.26344579585035, "violators_lhs": 7.9715818537821175, "violators_rhs": 4.712939467071792
"loss": 76.21790839912089, "violators_lhs": 7.973225262966167, "violators_rhs": 4.719670407457849
"loss": 139.63803117754136, "violators_lhs": 7.981806591022929, "violators_rhs": 4.710058326537539
"loss": 88.86843833160782, "violators_lhs": 7.981371936200952, "violators_rhs": 4.711409896055404
"loss": 18.282394593431473, "violators_lhs": 7.9855197850735395, "violators_rhs": 4.71678098778413
"loss": 55.311413324540844, "violators_lhs": 8.002361624532746, "violators_rhs": 4.703991787093649
"loss": 38.80947543350613, "violators_lhs": 8.004729458420092, "violators_rhs": 4.705664173265831
"loss": 212.34754931928455, "violators_lhs": 7.977859511282397, "violators_rhs": 4.699957776388723
"loss": 38.667435146222054, "violators_lhs": 7.992465983085718, "violators_rhs": 4.724521983185068
"loss": 93.41695229398844, "violators_lhs": 8.01700742224853, "violators_rhs": 4.717445388726296
"loss": 33.96085482659722, "violators_lhs": 7.980722023752851, "violators_rhs": 4.721812634794739
"loss": 64.01252438748274, "violators_lhs": 8.024392414652421, "violators_rhs": 4.7156260478285885
"loss": 495.0297211321344, "violators_lhs": 8.017785661358358, "violators_rhs": 4.724002467183561
"loss": 20.720368111041445, "violators_lhs": 8.029316432849969, "violators_rhs": 4.712811140410066
"loss": 37.64872062633174, "violators_lhs": 8.013159692181596, "violators_rhs": 4.717964904727802
"loss": 27.947737242612146, "violators_lhs": 8.025396260312704, "violators_rhs": 4.724439191790405
"loss": 43.52230365330491, "violators_lhs": 8.039845428466165, "violators_rhs": 4.731370901308518
"loss": 44.74899842928392, "violators_lhs": 8.02056538243415, "violators_rhs": 4.735427679646977
"loss": 31.898705942911032, "violators_lhs": 8.024835348613864, "violators_rhs": 4.738807638334071
```

**Comment by lw** [comment]
Since training is non-deterministic you can't expect to get the exact same answers as the ones we've put in the documentation as an example. Also, manual inspection isn't a great indicator of the quality of the model. What is the MRR that gets reported when you run the evaluation on your learned embeddings? If it's close to 0.79, then your embeddings are statistically as good as the ones we produced.

**Comment by adamlerer** [comment]
The fact that the loss jumps around so much looks suspicious to me though... @dany-nonstop are you running the latest version of the code from github, or are you pip installing PBG? We recently pushed a bug fix that sometimes led to these spikes in the loss.

**Comment by dany-nonstop** [comment]
> 
> 
> Since training is non-deterministic you can't expect to get the exact same answers as the ones we've put in the documentation as an example. Also, manual inspection isn't a great indicator of the quality of the model. What is the MRR that gets reported when you run the evaluation on your learned embeddings? If it's close to 0.79, then your embeddings are statistically as good as the ones we produced.

Hi, @lw thanks for the comment. I understand the randomness, but having France as the first hit *consistently* raised my suspicion. I'll double check MRR and report back.

**Comment by dany-nonstop** [comment]
> 
> 
> The fact that the loss jumps around so much looks suspicious to me though... @dany-nonstop are you running the latest version of the code from github, or are you pip installing PBG? We recently pushed a bug fix that sometimes led to these spikes in the loss.

HI, @adamlerer , thanks for the reminder. As I said earlier I was using default pip repo's PBG 1.0.0, but I'll try to clone the repo and report back.

**Comment by adamlerer** [comment]
Wait, what is `rank.py` here? Nearest neighbor search should always return the input entity as the "nearest" neighbor (because its distance is 0). Are you applying the `capital_city` relation type somewhere?

**Comment by dany-nonstop** [comment]
hi, @adamlerer  as stated earlier, this rank.py is more or less just a copy of the [document's code on ranking](https://torchbiggraph.readthedocs.io/en/latest/downstream_tasks.html#ranking), with a little change to the file path. Anyway attached please find the code. You see that I'm trying to find the capital of France (/m/0f8l9c), and the top result is always itself.

```
import json
import h5py
import torch
from torchbiggraph.model import ComplexDiagonalDynamicOperator, DotComparator

# Load entity count
with open("data/fb15k/entity_count_all_0.txt", "rt") as tf:
    entity_count = int(tf.read().strip())

# Load count of dynamic relations
with open("data/fb15k/dynamic_rel_count.txt", "rt") as tf:
    dynamic_rel_count = int(tf.read().strip())

# Load the operator's state dict
with h5py.File("model/fb15k/model.v45.h5", "r") as hf:
    operator_state_dict = {
        "real": torch.from_numpy(hf["model/relations/0/operator/rhs/real"][...]),
        "imag": torch.from_numpy(hf["model/relations/0/operator/rhs/imag"][...]),
    }
operator = ComplexDiagonalDynamicOperator(400, dynamic_rel_count)
operator.load_state_dict(operator_state_dict)
comparator = DotComparator()

# Load the offsets of the entities and the index of the relation type
with open("data/fb15k/entity_names_all_0.json", "rt") as tf:
    entity_names = json.load(tf)
src_entity_offset = entity_names.index("/m/0f8l9c")  # France
with open("data/fb15k/dynamic_rel_names.json", "rt") as tf:
    rel_type_names = json.load(tf)
rel_type_index = rel_type_names.index("/location/country/capital")

# Load the trained embeddings
with h5py.File("model/fb15k/embeddings_all_0.v45.h5", "r") as hf:
    src_embedding = torch.from_numpy(hf["embeddings"][src_entity_offset, :])
    dest_embeddings = torch.from_numpy(hf["embeddings"][...])

# Calculate the scores
scores, _, _ = comparator(
    comparator.prepare(src_embedding.view(1, 1, 400)).expand(1, entity_count, 400),
    comparator.prepare(
        operator(
            dest_embeddings,
            torch.tensor([rel_type_index]).expand(entity_count),
        ).view(1, entity_count, 400),
    ),
    torch.empty(1, 0, 400),  # Left-hand side negatives, not needed
    torch.empty(1, 0, 400),  # Right-hand side negatives, not needed
)

# Sort the entities by their score
permutation = scores.flatten().argsort(descending=True)
top5_entities = [entity_names[index] for index in permutation[:5]]

print(top5_entities)
```

**Comment by dany-nonstop** [comment]
> 
> 
> Since training is non-deterministic you can't expect to get the exact same answers as the ones we've put in the documentation as an example. Also, manual inspection isn't a great indicator of the quality of the model. What is the MRR that gets reported when you run the evaluation on your learned embeddings? If it's close to 0.79, then your embeddings are statistically as good as the ones we produced.

As suspected, the model doesn't converge. After 50 epochs, the evaluation results is pretty bad.

```
2020-05-20 13:12:51,458   [Evaluator] Starting edge path 1 / 1 (data/fb15k/test_partitioned)
2020-05-20 13:12:56,167   [Evaluator] ( 0 , 0 ): Processed 59071 edges in 4.7 s (0.013M/sec); load time: 0.0022 s
2020-05-20 13:12:56,167   [Evaluator] Stats for edge path 1 / 1, bucket ( 0 , 0 ): pos_rank:  1022.85 , mrr:  0.163422 , r1:  0.0816137 , r10:  0.337949 , r50:  0.540028 , auc:  0.930372 , count:  59071
2020-05-20 13:12:56,168   [Evaluator] 
2020-05-20 13:12:56,168   [Evaluator] Stats for edge path 1 / 1: pos_rank:  1022.85 , mrr:  0.163422 , r1:  0.0816137 , r10:  0.337949 , r50:  0.540028 , auc:  0.930372 , count:  59071
2020-05-20 13:12:56,168   [Evaluator] 
2020-05-20 13:12:56,168   [Evaluator] 
2020-05-20 13:12:56,168   [Evaluator] Stats: pos_rank:  1022.85 , mrr:  0.163422 , r1:  0.0816137 , r10:  0.337949 , r50:  0.540028 , auc:  0.930372 , count:  59071
2020-05-20 13:12:56,168   [Evaluator] 
```

**Comment by adamlerer** [comment]
Can you try just running the example script directly? E.g. `torchbiggraph_example_fb15k.py` or `python torchbiggraph/examples/fb15k.py`? I ran that and I'm getting MRR of 0.77.

And make sure you delete your `models/` folder first to make sure you don't have corrupted data.

**Comment by dany-nonstop** [comment]
hi, @adamlerer and @lw, I moved to a new directory and upgraded from 1.0.0 to torchbiggraph-1.0.1.dev0 by installing from the source code. Thanks for all the hints you gave me, I was able to get mrr=0.75 reported by `python torchbiggraph/examples/fb15k.py` on GPU training. snippet of the output
```
 [Evaluator] Stats: loss:  4.15954 , pos_rank:  45.2119 , **mrr:  0.750912** , r1:  0.655812 , r10:  0.887229 , r50:  0.937389 , auc:  0.991485 , count:  59071
```
So it looks like the training is converging now. However, the evaluation reports mrr of merely 0.53, is this something to be expected?
```
PyTorch-BigGraph$ torchbiggraph_eval torchbiggraph/examples/configs/fb15k_config_gpu.py -p edge_paths=data/$DATASET/test_partitioned
2020-05-20 14:33:18,152   [Evaluator] Starting edge path 1 / 1 (data/fb15k/test_partitioned)
2020-05-20 14:33:20,016   [Evaluator] ( 0 , 0 ): Processed 59071 edges in 1.8 s (0.032M/sec); load time: 0.031 s
2020-05-20 14:33:20,016   [Evaluator] Stats for edge path 1 / 1, bucket ( 0 , 0 ): loss:  4.79066 , pos_rank:  15.948 , mrr:  0.534513 , r1:  0.386264 , r10:  0.806157 , r50:  0.926588 , auc:  0.989546 , count:  59071
2020-05-20 14:33:20,016   [Evaluator] 
2020-05-20 14:33:20,016   [Evaluator] Stats for edge path 1 / 1: loss:  4.79066 , pos_rank:  15.948 , **mrr:  0.534513** , r1:  0.386264 , r10:  0.806157 , r50:  0.926588 , auc:  0.989546 , count:  59071
2020-05-20 14:33:20,016   [Evaluator] 
2020-05-20 14:33:20,016   [Evaluator] 
2020-05-20 14:33:20,016   [Evaluator] Stats: loss:  4.79066 , pos_rank:  15.948 , **mrr:  0.534513** , r1:  0.386264 , r10:  0.806157 , r50:  0.926588 , auc:  0.989546 , count:  59071
2020-05-20 14:33:20,016   [Evaluator]
```

**Comment by adamlerer** [comment]
Ah, for FB15k you have to do a special filtered eval with all negatives,
that's what you see from running `fb15k.py`. The result you see from just
running `torchbiggraph_eval` is our regular evaluation, which does not use
the fb15k evaluation procedure used in the literature.

**Comment by dany-nonstop** [comment]
got it, many thanks!
